### PR TITLE
[react-autosuggest] update signatures, documentation, use generic type

### PR DESCRIPTION
**Pull request by rosskevin**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/moroshko/react-autosuggest
- [x] Increase the version number in the header if appropriate. (same version)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

## Changes
- Convert Autosuggest component to a generic
- Update the `renderSuggestionsContainer` signature
- Add readme with usage example
- Document component props
- Export function types for easier reuse/passing in as props

### CODE REVIEW

**Comment by rosskevin** [comment]
From previous authors, I need a solid review on this.  

All tests succeed locally, but when I try to `yarn link` this package and use it in my project, I'm seeing 
```
Error:(106, 7) TS2605:JSX element type 'Autosuggest<Language>' is not a constructor function for JSX elements.
  Property 'render' is missing in type 'Autosuggest<Language>'.
```
I'm new to updating DefinitelyTyped, so I may be missing something.

Certainly though, `tsc` and `lint` runs on this PR, just not when I'm trying to use it in my project via `yarn link`.

**Comment by cdeutsch** [comment]
@rosskevin can you share your code where you try to use it?

Sounds like the `import` statement might be wrong

File: types/react-autosuggest/index.d.ts [hunk 1]
```diff
@@ -5,84 +5,200 @@
...
 //                 Terry Bayne <https://github.com/tbayne>
 //                 Christopher Deutsch <https://github.com/cdeutsch>
+//                 Kevin Ross <https://github.com/rosskevin>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.3
 
 import * as React from 'react';
 
-declare class Autosuggest extends React.Component<Autosuggest.AutosuggestProps> {}
+declare class Autosuggest<T> extends React.Component<Autosuggest.AutosuggestProps<T>> {}
```
Comment (cdeutsch) [inline types/react-autosuggest/index.d.ts]: Would changing this to `declare class Autosuggest<T = any>` make this backward compatible?

File: types/react-autosuggest/react-autosuggest-tests.tsx [hunk 2]
```diff
@@ -89,7 +91,7 @@ export class ReactAutosuggestBasicTest extends React.Component<any, any> {
             sectionTitle: { color: 'blue' }
         };
 
-        return <Autosuggest
+        return <LanguageAutosuggest
```
Comment (cdeutsch) [inline types/react-autosuggest/react-autosuggest-tests.tsx]: If possible, we should leave this as is to test backward compatibility and duplicate it to test the `LanguageAutosuggest` version

**Review by cdeutsch — CHANGES_REQUESTED** [review]

File: types/react-autosuggest/index.d.ts (hunk 1 above)
Comment (rosskevin) [reply types/react-autosuggest/index.d.ts]: Added, tests work.

File: types/react-autosuggest/react-autosuggest-tests.tsx (hunk 2 above)
Comment (rosskevin) [reply types/react-autosuggest/react-autosuggest-tests.tsx]: Left the `ReactAutosuggestBasicTest` for backward compatibility, added `ReactAutosuggestTypedTest` which is identical but strongly typed.

**Comment by rosskevin** [comment]
@cdeutsch regarding the imports I had already read that and I linked to the ultimate result of it in the README that I added.  It works fine in tests, so not sure why I have a problem with `yarn link`.    I'm inclined to think that my local setup is an environmental problem and that tests are correct, otherwise this would be a big problem for DefinitelyTyped.

File: types/react-autosuggest/index.d.ts [hunk 3]
```diff
@@ -5,84 +5,200 @@
...
+    type GetSectionSuggestions<TSuggestion> = (section: any) => TSuggestion[];
+    type GetSuggestionValue<TSuggestion> = (suggestion: TSuggestion) => string;
+    type OnSuggestionHighlighted = (params: SuggestionHighlightedParams) => void;
+    type SuggestionsFetchRequested = (request: SuggestionsFetchRequestedParams) => void;
+    type OnSuggestionsClearRequested = () => void;
+    type OnSuggestionSelected<TSuggestion> = (
+        event: React.FormEvent<any>,
+        data: SuggestionSelectedEventData<TSuggestion>,
+    ) => void;
+    type RenderInputComponent<TSuggestion> = (inputProps: InputProps<TSuggestion>) => JSX.Element;
```
Comment (in19farkt) [inline types/react-autosuggest/index.d.ts]: I would suggest it will be more correct to use `React.ReactNode` instead of `JSX.Element`.

**Review by cdeutsch — APPROVED** [review]
I ran `npm run lint react-autosuggest` on my machine, against your branch, and didn't get any errors.

Thanks for making those changes 👍

**Pull request merged**